Mathlib/Topology/GPseudoMetric/Basic.lean

```diff
@@ -0,0 +1,328 @@
+/-
+Copyright (c) 2024 Edward van de Meent. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Edward van de Meent
+-/
+import Mathlib.Algebra.CovariantAndContravariant
+import Mathlib.Algebra.NeZero
+import Mathlib.Algebra.Order.Monoid.Lemmas
+import Mathlib.Data.Set.Lattice
+/-!
+# General Pseudo-Metrics
```

Comment:
I'm wondering if this title can be made more specific. Is there a name for this concept in textbooks or papers? If there is, it could be useful to reference this.

To be clear: if you know this, it's wonderful - but in my personal opinion, this is not required to have your PR merged.

Reply:
<rant>
I did some digging, and according to [this mathoverflow post](https://mathoverflow.net/a/363831) among others, these are called `[Codomain]-Metrics` or `[Codomain]-valued metrics` (and the associated spaces `[Codomain]-Metric space`, although in my definition the requirements on the codomain is somewhat weaker than described in that post. if you'd like i can change the file to reflect this? also, in the post, both `𝑅` and `Λ` are used to refer to the codomain... since the latter might be confused with conjunction, i would prefer the first, but i have no clue how common either of these uses are... but then again, [this paper](https://arxiv.org/pdf/1502.05002.pdf) does refer to (something similar to) this concept as "generalised metric space". interestingly, all of these references put more conditions on the codomain than i expected... in particular, they expect the ordering to be co- and contravariant compared to addition...
</rant>
tldr: the version in the PR is close to an existing concept, which gets described as "generalised metric space". such spaces are called `R`-Metric spaces or `R`-valued metric spaces where `R` is the codomain, although capital Lambda is also sometimes used. The concept in this PR is slightly weaker with regards to the assumptions on the codomain.

Reply:
also, according to [this other paper](http://www.tac.mta.ca/tac/reprints/articles/1/tr1.pdf), there are ties with category theory as well